File: leetcode-cn-cpp/leetcode-cn-279.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int numSquares(int n) {
      if (n <= 1) return n;
      int* re = new int[n+1];
      re[0] = 0;
      for (int i=1; i<=n; i++) {
        int minNum = INT_MAX;
        for (int j=1; j*j<=i; j++) {
          minNum = min(minNum, re[i-j*j]);
        }
        re[i] = minNum+1;
      }
      int result = re[n];
      delete[] re;
      return result;
    }
};

int  main() {
  Solution s;
  cout<<s.numSquares(12)<<endl;
  return 0;
}

File: leetcode-cn-python/leetcode-cn-279.py
import sys
import re

class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """

        re = []
        re.append(0)
        for i in range(1, n+1):
            minNum = sys.maxint
            j = 1
            while j*j <= i:
                minNum = min(minNum, re[i - j*j])
                j = j+1
            re.append(minNum+1)
        print re
        return re[n]

if __name__ == "__main__":
    s = Solution()
    print s.numSquares(12)

    print re.match("tion", "function")
    print re.search("tion", "function")